### PR TITLE
Either instance for IsForeign

### PR DESCRIPTION
**Pull request by parsonsmatt**

### CODE REVIEW

File: src/Data/Foreign/Class.purs [hunk 1]
```diff
@@ -56,6 +56,15 @@ instance arrayIsForeign :: (IsForeign a) => IsForeign (Array a) where
     readElement :: forall a. (IsForeign a) => Int -> Foreign -> F (Array a)
     readElement i value = readWith (ErrorAtIndex i) value
 
+instance eitherIsForeign :: (IsForeign l, IsForeign r) => IsForeign (Either l r) where
+  read value =
+    case (read value) :: F r of
+         Right rval -> pure (Right rval)
```
Comment (paf31) [inline src/Data/Foreign/Class.purs]: So this prefers the `Right` parse then? I think `Left` would be more common (`aeson` etc.)

Also,  you can probably shorten this using `<|>`.

**Comment by paf31** [comment]
Thanks!

File: src/Data/Foreign/Class.purs (hunk 1 above)
Comment (parsonsmatt) [inline src/Data/Foreign/Class.purs]: Yup! My thinking is: Attempt to parse the thing we actually want, and if that fails, then attempt to parse the error case. If you have `Either String SomeType` then `String` would always succeed even if SomeType.

**Comment by parsonsmatt** [comment]
The `<|>` implementation is much nicer, thanks for the tip :+1:

**Comment by garyb** [comment]
I thought we weren't doing this? Previously: #26 #30

**Comment by paf31** [comment]
I think I would prefer to provide this as a function instead of an instance.

**Comment by parsonsmatt** [comment]
I've implemented two of the functions, one Right biased and the other Left biased, and removed the instance.

File: examples/Either.purs [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+instance pointIsForeign :: IsForeign Point where
+  read value = Point <$> readProp "x" value
+                     <*> readProp "y" value
+                     <*> readProp "z" value
+
+type Response = Either (Array String) Point
+
+main = do
+  print $
+    readJSON """{ "x":1, "y": 2, "z": 3}""" :: F Response
```
Comment (paf31) [inline examples/Either.purs]: How does this typecheck?

File: src/Data/Foreign/Class.purs [hunk 3]
```diff
@@ -75,3 +76,13 @@ readWith f value = either (Left <<< f) Right (read value)
 -- | Attempt to read a property of a foreign value at the specified index
 readProp :: forall a i. (IsForeign a, Index i) => i -> Foreign -> F a
 readProp prop value = value ! prop >>= readWith (errorAt prop)
+
+-- | Attempt to read a value that can be either one thing or another. This
+-- | implementation is right biased.
+readEitherR :: forall l r. (IsForeign l, IsForeign r) => Foreign -> F (Either l r)
```
Comment (paf31) [inline src/Data/Foreign/Class.purs]: Looks good!

File: examples/Either.purs (hunk 2 above)
Comment (parsonsmatt) [inline examples/Either.purs]: It didn't, and neither `psc-ide` nor `pulp -w build` compile the `examples` directory so I forgot to change it.

Comment (paf31) [inline examples/Either.purs]: Ah ok, thought I was missing something :smile:

**Pull request closed without merging**